NXP drivers: add pca9422 drivers(mfd, charger, and regulator)

taieup-kim opened the pull request:
Add drivers for NXP PCA9422 PMIC and it is for NXP MIMXRT700 board.

taieup-kim commented on include/zephyr/drivers/mfd/pca9422.h:
pca9422 has other functions like AMUX and it needs to add new driver for it later. To extend it, this header added to the public APIs.

taieup-kim replied:
I'm sorry that I cannot understand your meaning in detail.
In this header, it might be added some other definitions or parameter structures in the future.
Is it is different from other mfd header files?
How do you want to change it?

taieup-kim replied:
Most other mfd drivers also use public headers and use them only for "child" drivers.
I want to keep the same format with them.
If I follow your recommendation, I should add "zephyr_library_include_directories" for each child drivers.
I don't want it because it is different form from other child drivers.

taieup-kim commented on drivers/regulator/regulator_pca9422.c:
```diff
@@ -0,0 +1,846 @@
+					   PCA9422_BIT_SHIP_MODE, PCA9422_BIT_SHIP_MODE);
+}
+
+static const struct regulator_parent_driver_api parent_api = {
```
I will fix it.

dsseng commented on drivers/regulator/regulator_pca9422.c:
```diff
@@ -0,0 +1,846 @@
+	 * Fix DVS control to 01b, BUCKxOUT_DVS0 Register through I2C in ACTIVE mode,
+	 * and the voltage set to BUCKxOUT_SLEEP when in SLEEP mode
+	 */
+	reg_val = (1U << PCA9422_BIT_B1_DVS_CTRL_POS) | (1U << PCA9422_BIT_B2_DVS_CTRL_POS) |
+		  (1U << PCA9422_BIT_B3_DVS_CTRL_POS);
```
```suggestion
	reg_val = BIT(PCA9422_BIT_B1_DVS_CTRL_POS) | BIT(PCA9422_BIT_B2_DVS_CTRL_POS) |
		  BIT(PCA9422_BIT_B3_DVS_CTRL_POS);
```

dsseng commented on drivers/mfd/mfd_pca9422.c:
```diff
@@ -0,0 +1,402 @@
+#elif defined(CONFIG_SOC_SERIES_IMXRT7XX)
+	POWER_EnableInterrupts(PMC_INTRCTRL_INTNIE_MASK);
+#else
+	/* Enable GPIO interrupt */
```
Does this need an else clause at all?

dsseng commented on drivers/charger/charger_pca9422.c:
```diff
@@ -0,0 +1,672 @@
+	uint8_t num_ranges;
+	int ret;
+
+	voltage_uv = CLAMP(voltage_uv, VBAT_REG_MIN_UV, VBAT_REG_MAX_UV);
```
Should out of range values trigger EINVAL or ENOTSUP to alert the app developer? I'm not a maintainer, but had this thought. Some of the subsys drivers return errors, some do clamp to the range, I do not know the idiomatic way

decsny reviewed (changes requested):
needs clarifications

taieup-kim commented on drivers/charger/charger_pca9422.c:
```diff
@@ -0,0 +1,664 @@
+	}
+
+	/* Set interrupt handler */
+	child_idx = DT_NODE_CHILD_IDX(DT_NODELABEL(charger));
```
It is a bug.
child_idx is not match to pca9422 device enum type in the current code.
My original plan is to make sub level interrupt handler for each child device.
Currently only charger driver has sub level interrupt handler and doesn't need child_idx as you mentioned.
However, if other driver like ADC is implemented later, child_idx might be needed.
I will change this instance structure.

taieup-kim commented on drivers/charger/charger_pca9422.c:
```diff
@@ -0,0 +1,664 @@
+                                                                                                   \
+	DEVICE_DT_INST_DEFINE(inst, &pca9422_charger_init, NULL, &charger_pca9422_data_##inst,     \
+			      &charger_pca9422_config_##inst, POST_KERNEL,                         \
+			      CONFIG_MFD_INIT_PRIORITY, &pca9422_charger_driver_api);
```
I set charger driver as the same init priority of MFD and want to initialize earlier than other device drivers.
If a customer powers up with TA, we need to initialize charging parameter earlier than others because it is related to system power.

taieup-kim replied:
MFD device checks chip ID and revision, so it should be initialized first.
Init charger priority number is higher than mfd number, so mfd device initialization is earlier than charger device.
PCA9422 charger driver is one of child devices of PCA9422 MFD driver and it already defined in DT, so it already considers dependency for each child devices.
I think that there is no error for this priority change.

taieup-kim commented on drivers/mfd/mfd_pca9422.c:
```diff
@@ -0,0 +1,397 @@
+
+struct mfd_pca9422_data {
+	struct k_work work;
+	struct k_sem lock;
```
If we don't use semaphore and I2C operation happens in other child driver, register value can be changed.
Even though I disable the interrupt, register values can be changed.
If other driver processes interrupt event of PCA9422 like TA plug-in or out, it can set charger control register of PCA9422.
At this time, if other driver set wrong register value, charger also might be wrong behavior.
I want to block other i2c operation when reading interrupt registers.

decsny replied:
This is not a high bandwidth use case, it's a regulator. The context is that they are doing a few synchronous i2c transfers (not async). They put a semaphore to guard this "critical section" of using the i2c bus, and my question is if the i2c synchronous API already blocks the caller until the bus is available , so that we don't need this semaphore in this driver.

teburd replied:
> This is not a high bandwidth use case, it's a regulator. The context is that they are doing a few synchronous i2c transfers (not async). They put a semaphore to guard this "critical section" of using the i2c bus, and my question is if the i2c synchronous API already blocks the caller until the bus is available , so that we don't need this semaphore in this driver.

Then why are you bringing up the async APIs? High bandwidth isn't always a goal with async calls, at times its to save precious memory from wasteful stacks.

decsny replied:
> > This is not a high bandwidth use case, it's a regulator. The context is that they are doing a few synchronous i2c transfers (not async). They put a semaphore to guard this "critical section" of using the i2c bus, and my question is if the i2c synchronous API already blocks the caller until the bus is available , so that we don't need this semaphore in this driver.
> 
> Then why are you bringing up the async APIs? High bandwidth isn't always a goal with async calls, at times its to save precious memory from wasteful stacks.

I'm not sure where you think I was focusing on async calls. All I was asking is if the client code using any of the i2c APIs needs to synchronize the bus access or if it is already done by the i2c bus controller drivers.

taieup-kim replied:
For the single I2C operation, it already has semaphore in its I2C API and it is synchronous API.
My original goal for this semaphore is not i2c operation issue. It is not  synchronous or asynchronous problem.
I assumed that if context switching happens when MFD driver get the interrupt and read some registers for interrupts during charging operation working, it might make unexpected charging operation. So, I wanted to prevent unexpected context switching during reading interrupt registers.
However, I checked it again and confirmed charger parameters cannot change until reading interrupt registers are done. So I removed it.

rriveramcrus commented on drivers/charger/charger_pca9422.c:
```diff
@@ -0,0 +1,664 @@
+		break;
+
+	default:
+		ret = -ENOTSUP;
+		break;
```
could just return here, the `break` is not necessary.

rriveramcrus commented on drivers/charger/charger_pca9422.c:
```diff
@@ -0,0 +1,664 @@
+	int ret = 0;
+
+	switch (prop) {
+	case CHARGER_PROP_ONLINE:
+		ret = pca9422_charger_get_online(dev, &data->online);
+		val->online = data->online;
+		break;
+
+	case CHARGER_PROP_STATUS:
+		ret = pca9422_charger_get_status(dev, &data->status);
+		val->status = data->status;
+		break;
```
Shouldn't we check the `ret` before writing to `data`?

decsny commented on drivers/mfd/mfd_pca9422.c:
```diff
@@ -0,0 +1,397 @@
+	};                                                                                         \
+                                                                                                   \
+	static struct mfd_pca9422_data mfd_pca9422_data_##inst = {                                 \
+		.work = Z_WORK_INITIALIZER(mfd_pca9422_work_handler),                              \
+		.lock = Z_SEM_INITIALIZER(mfd_pca9422_data_##inst.lock, 1, 1),                     \
```
I don't care, these macros are not allowed to be used, don't follow a bad example. use the public APIs K_SEM_DEFINE etc

decsny commented on drivers/mfd/Kconfig.pca9422:
```diff
@@ -0,0 +1,10 @@
+# Copyright 2025 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+config MFD_PCA9422
```
That is not an excuse to not document what is going on here. As far as I'm concerned it's the same hardware that you are introducing two different driver for, the explanation of what is going on here needs to be clear. Your claim that they don't affect each others state needs to be explained because clearly they are affecting the state of the same hardware. Are they supposed to be used at the same time or not?

taieup-kim commented on drivers/mfd/mfd_pca9422.c:
```diff
@@ -0,0 +1,397 @@
+	 * instead of GPIO interrupt, should disable PMIC pin interrupt here
+	 * before clearing PMIC_IRQ flag
+	 */
+#if defined(CONFIG_SOC_SERIES_IMXRT5XX)
+	POWER_DisableInterrupts(PMC_CTRL_INTRPADEN_MASK);
+#elif defined(CONFIG_SOC_SERIES_IMXRT7XX)
+	POWER_DisableInterrupts(PMC_INTRCTRL_INTNIE_MASK);
```
I will implement them as __weak function.

taieup-kim commented on drivers/mfd/mfd_pca9422.c:
```diff
@@ -0,0 +1,397 @@
+	struct k_sem lock;
+	const struct device *dev;
+	struct mfd_pca9422_child *children;
+	size_t num_children;
```
I will remove it.

kartben commented on include/zephyr/drivers/mfd/pca9422.h:
```diff
@@ -0,0 +1,109 @@
+
+typedef void (*child_isr_t)(const struct device *dev);
+
+enum {
+	PCA9422_DEV_REG,
+	PCA9422_DEV_CHG,
+	PCA9422_DEV_FG,
+	PCA9422_DEV_MAX,
+};
+
+/**
+ * @brief Set child interrupt handler of pca9422
+ *
+ * @param dev pca9422 mfd device
+ * @param child_dev pca9422 child device
+ * @param child_idx index in enum type of child device
+ * @param handler interrupt handler of child device
+ */
+void mfd_pca9422_set_irqhandler(const struct device *dev, const struct device *child_dev,
+				uint8_t child_idx, child_isr_t handler);
```
give the enum a name and actually use it instead of the quite obscure "uint8_t child_index"? also, the enum would need doxygen comments if it's meant to be API

taieup-kim replied:
I will change child device enum type name and use it instead of uint8_t.

taieup-kim committed:
drivers: mfd: add pca9422 drivers(mfd+charger+regulator)

Add drivers for NXP PCA9422 PMIC that includes charger and regulators.

Signed-off-by: Clark Kim <clark.kim@nxp.com>

taieup-kim committed:
board: nxp: mimxrt700_evk: add pmic instance for rt700 cm33_cpu0

enable pmic i2c clock
config pmic pin
add pca9422 pmic instance on rt700_evk cm33_cpu0

Signed-off-by: Clark Kim <clark.kim@nxp.com>

taieup-kim committed:
soc: nxp: imxrt7xx: Add pmic interrupt APIs

Add pmic interrupt enable/disable/clear APIs

Signed-off-by: Clark Kim <clark.kim@nxp.com>

rriveramcrus commented:
Hoping to get my final review in by Thursday, but overall looking good.

rriveramcrus commented on drivers/charger/charger_pca9422.c:
```diff
@@ -0,0 +1,654 @@
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/sys/linear_range.h>
+#include "zephyr/logging/log.h"
```
This is supposed to be `<zephyr/logging/log.h>`

taieup-kim replied:
I will change it.

rriveramcrus commented on drivers/charger/charger_pca9422.c:
```diff
@@ -0,0 +1,654 @@
+	}
+
+	/* Unlock charger control register */
+	val = PCA9422_CHAREGER_UNLOCK;
```
Must fix: should be `PCA9422_CHARGER_UNLOCK

Nit:
Does this safe write mechanism need to be locked after initialization? Should we unlock then lock again in the `set_property` handler? Seems like a nice safety feature to make use of.

This can wait for a subsequent PR, but I withhold my approval until we've got a plan one way or the other.

rriveramcrus reviewed (changes requested):
@taieup-kim, just some minor stuff to address and then I'm good to approve.

taieup-kim commented on drivers/charger/charger_pca9422.c:
```diff
@@ -0,0 +1,654 @@
+	ret = mfd_pca9422_reg_burst_read(config->mfd, PCA9422_REG_INT_DEVICE_0, int_val, 6);
+	if (ret == 0) {
+		/* Check interrupt event */
+		LOG_INF("%s: int_device[0]=0x%x,  [1]=0x%x\n", __func__, int_val[0], int_val[1]);
```
I will change LOG_INF to LOG_DBG.

taieup-kim commented on drivers/charger/charger_pca9422.c:
```diff
@@ -0,0 +1,654 @@
+	default:
+		return -ENOTSUP;
+	}
+	return ret;
```
I will remove "return ret;" in line 514. Do not need it.

taieup-kim commented:
@rriveramcrus 
I applied your change request. Can you approve for new commit?

Can any other reviewers review and approve for it?

rriveramcrus reviewed (approved):
Thank you for your contribution!
